File: First Bad Version - Leetcode 278/First Bad Version - Leetcode 278.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int i,j,k,c=0;
        int h=n,l=0,m;

        while(l<=h)
        {
            m=l+(h-l)/2;
            int res=isBadVersion(m);
            if(res==1 and (m==0 or isBadVersion(m-1)!=1)){
                return m;
            }
            else if(res==0){
                l=m+1;
            }else
            h=m-1;
        }
        return m;
        
    }
};

File: First Bad Version - Leetcode 278/First Bad Version - Leetcode 278.py
# Brute Force Solution
class Solution:
    def firstBadVersion(self, n: int) -> int:
        for version in range(1, n+1):
            if isBadVersion(version):
                return version
    # Time: O(n)
    # Space: O(1)

# Optimal Solution
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:
class Solution:
    def firstBadVersion(self, n: int) -> int:
        L = 1
        R = n

        while L < R:
            M = (L+R) // 2
            if isBadVersion(M):
                R = M
            else:
                L = M + 1
        
        return L 
        # Time: O(Log n)
        # Space: O(1)
